File: typescript/src/messages/vaas_verdict.ts
import { JsonProperty, JsonObject } from "typescript-json-serializer";
import { Verdict } from "../Verdict";
import { FileReport } from "./FileReport";
import { UrlReport } from "./UrlReport";

@JsonObject()
export class VaasVerdict {
  public constructor(
    @JsonProperty() public sha256: string,
    @JsonProperty() public verdict: Verdict,
    @JsonProperty() public detection: string | undefined,
    @JsonProperty({ name: "fileType" }) public file_type: string | undefined,
    @JsonProperty({ name: "mimeType" }) public mime_type: string | undefined,
    @JsonProperty({ name: "isEncrypted" })
    public is_encrypted: boolean | undefined,
  ) {}

  public static fromFileReport(report: FileReport): VaasVerdict {
    return new VaasVerdict(
      report.sha256,
      report.verdict,
      report.detection,
      report.file_type,
      report.mime_type,
      report.is_encrypted,
    );
  }

  public static fromUrlReport(report: UrlReport): VaasVerdict {
    return new VaasVerdict(
      report.sha256,
      report.verdict,
      report.detection,
      report.file_type,
      report.mime_type,
      report.is_encrypted,
    );
  }
}

File: rust/src/vaas_verdict.rs
//! # VaaS Verdict
//!
//! The `VaaSVerdict` is the result of a request for a verdict. It contains the analysis results from the server.

use crate::message::report::{FileReport, UrlReport};
use crate::message::verdict::Verdict;
use crate::sha256::Sha256;

/// Response with all information regarding an analysis.
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct VaasVerdict {
    /// Sha256 of the requested file
    pub sha256: Sha256,
    /// Verdict for the file
    pub verdict: Verdict,
    /// The detected malware or PUP found in the file, if any
    pub detection: Option<String>,
    /// File type as classified by https://www.darwinsys.com/file/
    pub file_type: Option<String>,
    /// mime type as classified by https://www.darwinsys.com/file/
    pub mime_type: Option<String>,
}

impl From<FileReport> for VaasVerdict {
    fn from(report: FileReport) -> Self {
        Self {
            sha256: report.sha256,
            verdict: report.verdict,
            detection: report.detection,
            file_type: report.file_type,
            mime_type: report.mime_type,
        }
    }
}

impl From<UrlReport> for VaasVerdict {
    fn from(report: UrlReport) -> Self {
        Self {
            sha256: report.sha256,
            verdict: report.verdict,
            detection: report.detection,
            file_type: report.file_type,
            mime_type: report.mime_type,
        }
    }
}
